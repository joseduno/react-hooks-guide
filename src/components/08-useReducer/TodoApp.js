import React, { useEffect, useReducer } from 'react'
import AddTodo from './AddTodo'
import TodoList from './TodoList'
import { todoReducer } from './todoReducer'

// Con el local storage, puedo mantener valores aunque la pagina se actualice completamente
const init = () => JSON.parse(localStorage.getItem('todos')) || []
// se obtiene el mismo resultado si en lugar de dfinirlo como una funcion,
// se define como un objeto y se pasa como segundo argumento al useReducer

const TodoApp = () => {

    // init funcion que se ejecuta por primera vez, [] valor inicial del stado del reducer
    // todoReducer es la funcion reducer, todos el estado del reducer, dispatch es la funcion
    // que permite ejecutar los diferentes acciones definidas en el reducer (todoReducer)
    const [todos, dispatch] = useReducer(todoReducer, [], init)
    
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleDelete = id => {
        const action = {
            type: 'delete',
            payload: id
        }
        dispatch(action)
    }

    const handleDone = id => {
        dispatch({
            type: 'toggle',
            payload: id
        })
    }

    const handleAdd = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        })
    }

    return (
        <div className='container mt-5'>
            <h1>TodoApp ({todos.length})</h1>
            <p>
                Pequeña app, muestra claramente que son y como funcionan las funciones 
                'reducers' (estados y acciones para manipularlos) y el uso del hook
                'useReducer' que constituyen las bases para entender Redux).
                Básicamente el useReducer se asemeja al useState, pero es más robusto, 
                ya que permite modificar el estado de muchas formas, más fácil de lo 
                se lograría utilizando netamente useState. Recomendación: utilizar useState 
                para cosas más simples, y useReducer para cosas más complejas
                <br /><br />
                También se muestra como utilizar el local storage del navegador en conjunto 
                con el useEffect
            </p>
            <div className='row'>
                <div className='col-7'>
                    <TodoList
                        todos={todos}
                        handleDelete={handleDelete}
                        handleDone={handleDone}
                    />
                </div>
                <div className='col mt-2'>
                    <AddTodo handleAdd = {handleAdd}/>
                </div>
            </div>
            <hr />
        </div>
    )
}

export default TodoApp
