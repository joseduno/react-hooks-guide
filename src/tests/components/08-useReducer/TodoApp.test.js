import { shallow } from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp";


describe('Pruebas en el compoente <TodoApp />', () => {

    const wrapper = shallow(<TodoApp />)

    // Para emular llamado a guardar en el local storage
    Storage.prototype.setItem = jest.fn(() => {})
    // el callback es para indicar el retorno ficticio de la jest function

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    })

    /* Comentado porque no funciona el mopunt en la version 17 de react
    test('debe agregar un TODO', () => {
        // Buscar diferencias entre mount (avanzado) y shallow (basico)

        const wrapper = mount(<TodoApp />)

        act(()=> {
            wrapper.find('AddTodo').prop('handleAdd')(demoTodos[0])
            wrapper.find('AddTodo').prop('handleAdd')(demoTodos[1])
        })


        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)')

        // Referencia al localstorage para saber si fue llamado
        expect(localStorage.setItem).toHaveBeenCalledTimes(2)

    })

    // Lo mismo para eliminar un TODO
     */

    
})
