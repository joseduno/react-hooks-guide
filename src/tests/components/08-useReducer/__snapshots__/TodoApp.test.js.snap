// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el compoente <TodoApp /> debe mostrarse correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <p>
    Pequeña app, muestra claramente que son y como funcionan las funciones 'reducers' (estados y acciones para manipularlos) y el uso del hook 'useReducer' que constituyen las bases para entender Redux). Básicamente el useReducer se asemeja al useState, pero es más robusto, ya que permite modificar el estado de muchas formas, más fácil de lo se lograría utilizando netamente useState. Recomendación: utilizar useState para cosas más simples, y useReducer para cosas más complejas
    <br />
    <br />
    También se muestra como utilizar el local storage del navegador en conjunto con el useEffect
  </p>
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleDone={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col mt-2"
    >
      <AddTodo
        handleAdd={[Function]}
      />
    </div>
  </div>
  <hr />
</div>
`;
